ceph: add labels to monitoring spec

umangachapagain opened the pull request:
Signed-off-by: Umanga Chapagain <chapagainumanga@gmail.com>

<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/master/development-flow.html)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->

**Description of your changes:**
Labels field in MonitoringSpec is used define labels which
needs to be injected into the monitoring resources created
by Rook operator.

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.

leseb reviewed (changes requested):
Please make sure your branch is up to date with `master` and then run `make crds-gen`

umangachapagain commented:
> Please make sure your branch is up to date with `master` and then run `make crds-gen`

Tried it. No diff.
I am on Go 1.16. Could that be an issue? (I hope not)

leseb commented:
> > Please make sure your branch is up to date with `master` and then run `make crds-gen`
> 
> Tried it. No diff.
> I am on Go 1.16. Could that be an issue? (I hope not)

Weird, the helm diff should not be that big.

leseb commented:
@umangachapagain which `yq` version do you use on your system?

umangachapagain commented:
> @umangachapagain which `yq` version do you use on your system?

jq 1.6

leseb commented:
> > @umangachapagain which `yq` version do you use on your system?
> 
> jq 1.6

`yq` not `jq` :-)

umangachapagain commented:
> > > @umangachapagain which `yq` version do you use on your system?
> > 
> > 
> > jq 1.6
> 
> `yq` not `jq` :-)

yq version 2.4.0

subhamkrai commented:
> > > > @umangachapagain which `yq` version do you use on your system?
> > > 
> > > 
> > > jq 1.6
> > 
> > 
> > `yq` not `jq` :-)
> 
> yq version 2.4.0

@umangachapagain try rebasing...recentally we have few changes in crds generation. We replaced `the command make crds-gen` to `make crds` and also we are pulling yq v3 directly

umangachapagain commented:
> 
> @umangachapagain try rebasing...recently we have few changes in crds generation. We replaced `the command make crds-gen` to `make crds` and also we are pulling yq v3 directly
>
Updated yq to v4. Rebased and re-generated everything.
Looks good now.

leseb commented on pkg/operator/ceph/cluster/mgr/mgr.go:
```diff
@@ -435,13 +436,18 @@ func wellKnownModule(name string) bool {
 	}
 	serviceMonitor.SetName(AppName)
 	serviceMonitor.SetNamespace(c.clusterInfo.Namespace)
+	err = mergo.Merge(serviceMonitor.ObjectMeta.Labels, labels, mergo.WithOverride)
```
Why using this package for merging?

umangachapagain commented:
/hold seeing some runtime issues

leseb reviewed (changes requested):
Please update this page https://rook.io/docs/rook/v1.5/ceph-monitoring.html

umangachapagain commented on cluster/examples/kubernetes/ceph/cluster.yaml:
```diff
@@ -78,6 +78,10 @@ spec:
     rulesNamespace: rook-ceph
+    # labels is a list of key-value pairs. It is injected into all the monitoring resources created by operator.
+    # These labels can be passed as LabelSelector to Prometheus
+    # labels:
```
I didn't know we had such spec already. Added "monitoring" to the section.

umangachapagain commented on pkg/operator/ceph/cluster/mgr/mgr.go:
```diff
@@ -435,13 +436,18 @@ func wellKnownModule(name string) bool {
 	}
 	serviceMonitor.SetName(AppName)
 	serviceMonitor.SetNamespace(c.clusterInfo.Namespace)
+	err = mergo.Merge(&serviceMonitor.ObjectMeta.Labels, labels, mergo.WithOverride)
```
Made the change. But, `ApplyToObjectMeta()` does not merge existing keys, it only adds new ones.
We'll eventually need to add a way to override existing labels.
